File: brique-104/php-sdk/examples/webhook_receiver.php
<?php

/**
 * Molam Webhook Receiver Example
 *
 * Secure webhook endpoint for receiving Molam events.
 * Verifies HMAC signatures and processes events.
 *
 * Usage:
 *   1. Set environment variables:
 *      export MOLAM_WEBHOOK_SECRET="whsec_..."
 *
 *   2. Run server:
 *      php -S localhost:8081 examples/webhook_receiver.php
 *
 *   3. Configure webhook URL in Molam dashboard:
 *      https://your-domain.com/webhook/molam
 *
 *   4. Test with curl:
 *      curl -X POST http://localhost:8081/webhook/molam \
 *        -H "Molam-Signature: t=1234567890000,v1=abc...,kid=v1" \
 *        -d '{"type":"payment_intent.succeeded","data":{...}}'
 */

require_once __DIR__ . '/../vendor/autoload.php';

use Molam\SDK\Client;
use Molam\SDK\Config;
use Molam\SDK\Utils\WebhookVerifier;

// Initialize SDK client
$config = new Config([
    'api_key' => getenv('MOLAM_API_KEY') ?: 'sk_test_...',
    'webhook_secret' => getenv('MOLAM_WEBHOOK_SECRET') ?: 'whsec_test_secret',
]);

$molam = new Client($config);

// Database connection (for storing webhook events)
// Adjust credentials as needed
$pdo = new PDO(
    'mysql:host=localhost;dbname=molam_app',
    getenv('DB_USER') ?: 'root',
    getenv('DB_PASS') ?: '',
    [PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION]
);

// Simple router
$method = $_SERVER['REQUEST_METHOD'];
$path = parse_url($_SERVER['REQUEST_URI'], PHP_URL_PATH);

header('Content-Type: application/json');

try {
    // Route: POST /webhook/molam
    if ($method === 'POST' && $path === '/webhook/molam') {
        // Get request data
        $signature = $_SERVER['HTTP_MOLAM_SIGNATURE'] ?? '';
        $payload = file_get_contents('php://input');

        // Verify signature
        try {
            $isValid = WebhookVerifier::verify(
                $signature,
                $payload,
                fn(string $kid) => $config->webhookSecret
            );

            if (!$isValid) {
                http_response_code(401);
                echo json_encode(['error' => 'Invalid signature']);
                exit;
            }
        } catch (InvalidArgumentException $e) {
            http_response_code(401);
            echo json_encode(['error' => $e->getMessage()]);
            exit;
        }

        // Parse event
        $event = json_decode($payload, true);

        // Store event in database (for processing queue)
        $stmt = $pdo->prepare(
            "INSERT INTO molam_webhook_events
            (event_id, event_type, payload, signature, verified)
            VALUES (:event_id, :event_type, :payload, :signature, :verified)"
        );
        $stmt->execute([
            'event_id' => $event['id'] ?? uniqid('evt_'),
            'event_type' => $event['type'] ?? 'unknown',
            'payload' => $payload,
            'signature' => $signature,
            'verified' => 1,
        ]);

        // Process event synchronously (or queue for async processing)
        processWebhookEvent($event, $molam, $pdo);

        // Return 200 OK
        http_response_code(200);
        echo json_encode(['status' => 'ok']);
        exit;
    }

    // Route: GET / (webhook info)
    if ($method === 'GET' && $path === '/') {
        echo json_encode([
            'name' => 'Molam Webhook Receiver',
            'version' => '1.0.0',
            'endpoint' => '/webhook/molam',
            'status' => 'active',
        ], JSON_PRETTY_PRINT);
        exit;
    }

    // 404 Not Found
    http_response_code(404);
    echo json_encode(['error' => 'Not found']);
} catch (Exception $e) {
    error_log('Webhook processing error: ' . $e->getMessage());

    http_response_code(500);
    echo json_encode([
        'error' => 'Internal server error',
        'message' => $e->getMessage(),
    ]);
}

/**
 * Process webhook event.
 *
 * @param array<string, mixed> $event Webhook event data
 * @param Client $molam Molam SDK client
 * @param PDO $pdo Database connection
 */
function processWebhookEvent(array $event, Client $molam, PDO $pdo): void
{
    $eventType = $event['type'] ?? 'unknown';
    $eventData = $event['data'] ?? [];

    error_log("Processing webhook event: {$eventType}");

    switch ($eventType) {
        case 'payment_intent.succeeded':
            handlePaymentSucceeded($eventData, $molam, $pdo);
            break;

        case 'payment_intent.failed':
            handlePaymentFailed($eventData, $molam, $pdo);
            break;

        case 'payment_intent.canceled':
            handlePaymentCanceled($eventData, $molam, $pdo);
            break;

        case 'refund.created':
            handleRefundCreated($eventData, $molam, $pdo);
            break;

        case 'refund.succeeded':
            handleRefundSucceeded($eventData, $molam, $pdo);
            break;

        case 'payout.paid':
            handlePayoutPaid($eventData, $molam, $pdo);
            break;

        default:
            error_log("Unknown event type: {$eventType}");
    }

    // Mark event as processed
    $stmt = $pdo->prepare(
        "UPDATE molam_webhook_events
        SET processed = 1, processed_at = NOW()
        WHERE event_id = :event_id"
    );
    $stmt->execute(['event_id' => $event['id'] ?? '']);
}

/**
 * Handle successful payment.
 */
function handlePaymentSucceeded(array $data, Client $molam, PDO $pdo): void
{
    $paymentId = $data['id'] ?? '';
    $amount = $data['amount'] ?? 0;
    $currency = $data['currency'] ?? 'USD';
    $orderId = $data['metadata']['order_id'] ?? null;

    error_log("Payment succeeded: {$paymentId} ({$amount} {$currency})");

    // Update order status in database
    if ($orderId) {
        $stmt = $pdo->prepare(
            "UPDATE orders SET status = 'paid', payment_id = :payment_id WHERE id = :order_id"
        );
        $stmt->execute([
            'payment_id' => $paymentId,
            'order_id' => $orderId,
        ]);
    }

    // Send confirmation email
    // sendPaymentConfirmationEmail($data);

    // Trigger fulfillment
    // triggerOrderFulfillment($orderId);
}

/**
 * Handle failed payment.
 */
function handlePaymentFailed(array $data, Client $molam, PDO $pdo): void
{
    $paymentId = $data['id'] ?? '';
    $errorCode = $data['error']['code'] ?? 'unknown';
    $orderId = $data['metadata']['order_id'] ?? null;

    error_log("Payment failed: {$paymentId} (error: {$errorCode})");

    // Update order status
    if ($orderId) {
        $stmt = $pdo->prepare(
            "UPDATE orders SET status = 'payment_failed', error = :error WHERE id = :order_id"
        );
        $stmt->execute([
            'error' => $errorCode,
            'order_id' => $orderId,
        ]);
    }

    // Send failure notification
    // sendPaymentFailureEmail($data);
}

/**
 * Handle canceled payment.
 */
function handlePaymentCanceled(array $data, Client $molam, PDO $pdo): void
{
    $paymentId = $data['id'] ?? '';
    $orderId = $data['metadata']['order_id'] ?? null;

    error_log("Payment canceled: {$paymentId}");

    // Update order status
    if ($orderId) {
        $stmt = $pdo->prepare(
            "UPDATE orders SET status = 'canceled' WHERE id = :order_id"
        );
        $stmt->execute(['order_id' => $orderId]);
    }
}

/**
 * Handle refund created.
 */
function handleRefundCreated(array $data, Client $molam, PDO $pdo): void
{
    $refundId = $data['id'] ?? '';
    $paymentId = $data['payment_id'] ?? '';
    $amount = $data['amount'] ?? 0;

    error_log("Refund created: {$refundId} for payment {$paymentId} ({$amount})");

    // Store refund record
    $stmt = $pdo->prepare(
        "INSERT INTO refunds (refund_id, payment_id, amount, status)
        VALUES (:refund_id, :payment_id, :amount, 'pending')"
    );
    $stmt->execute([
        'refund_id' => $refundId,
        'payment_id' => $paymentId,
        'amount' => $amount,
    ]);
}

/**
 * Handle successful refund.
 */
function handleRefundSucceeded(array $data, Client $molam, PDO $pdo): void
{
    $refundId = $data['id'] ?? '';
    $paymentId = $data['payment_id'] ?? '';

    error_log("Refund succeeded: {$refundId}");

    // Update refund status
    $stmt = $pdo->prepare(
        "UPDATE refunds SET status = 'succeeded', processed_at = NOW()
        WHERE refund_id = :refund_id"
    );
    $stmt->execute(['refund_id' => $refundId]);

    // Send refund confirmation email
    // sendRefundConfirmationEmail($data);
}

/**
 * Handle payout paid.
 */
function handlePayoutPaid(array $data, Client $molam, PDO $pdo): void
{
    $payoutId = $data['id'] ?? '';
    $amount = $data['amount'] ?? 0;
    $merchantId = $data['merchant_id'] ?? '';

    error_log("Payout paid: {$payoutId} to merchant {$merchantId} ({$amount})");

    // Update payout record
    $stmt = $pdo->prepare(
        "UPDATE payouts SET status = 'paid', paid_at = NOW()
        WHERE payout_id = :payout_id"
    );
    $stmt->execute(['payout_id' => $payoutId]);

    // Send payout notification
    // sendPayoutNotificationEmail($merchantId, $data);
}

File: brique-105/examples/webhook_receiver.py
"""
Webhook receiver example.

Demonstrates webhook signature verification and event processing.

Usage:
    # Standalone verification
    echo '{"event":"payment.succeeded"}' | python examples/webhook_receiver.py "t=1705420800000,v1=abc...,kid=1"

    # Flask webhook endpoint
    export MOLAM_WEBHOOK_SECRET="whsec_..."
    python examples/webhook_receiver.py --server
"""

import sys
import os
import json

# Add parent directory to path
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from molam_sdk.utils.webhook import verify_signature, parse_signature_header
from molam_sdk.exceptions import SignatureError


def get_secret_by_kid(kid: str) -> str:
    """
    Retrieve webhook secret by key ID.

    In production, this should fetch from:
    - Environment variables (for simple setups)
    - Database (for multi-tenant)
    - Vault/KMS (for high security)

    Args:
        kid: Key ID from signature header

    Returns:
        Webhook secret for the given key ID
    """
    # Simple implementation: single secret from environment
    # For multi-version support, add logic to handle multiple kids
    secret = os.getenv("MOLAM_WEBHOOK_SECRET", "")

    if not secret:
        raise SignatureError(f"No secret configured for kid={kid}")

    return secret


def process_webhook_event(event: dict) -> None:
    """
    Process webhook event.

    Args:
        event: Parsed webhook event data
    """
    event_type = event.get("type", "unknown")
    event_data = event.get("data", {})

    print(f"Processing event: {event_type}")

    # Handle different event types
    if event_type == "payment_intent.succeeded":
        handle_payment_succeeded(event_data)
    elif event_type == "payment_intent.failed":
        handle_payment_failed(event_data)
    elif event_type == "payment_intent.canceled":
        handle_payment_canceled(event_data)
    elif event_type == "refund.created":
        handle_refund_created(event_data)
    elif event_type == "refund.succeeded":
        handle_refund_succeeded(event_data)
    elif event_type == "payout.paid":
        handle_payout_paid(event_data)
    else:
        print(f"⚠ Unknown event type: {event_type}")


def handle_payment_succeeded(data: dict) -> None:
    """Handle successful payment event."""
    payment_id = data.get("id")
    amount = data.get("amount")
    currency = data.get("currency")
    metadata = data.get("metadata", {})
    order_id = metadata.get("order_id")

    print(f"✓ Payment succeeded: {payment_id}")
    print(f"  Amount: {amount} {currency}")
    print(f"  Order: {order_id}")

    # TODO: Update order status in database
    # TODO: Send confirmation email
    # TODO: Trigger fulfillment


def handle_payment_failed(data: dict) -> None:
    """Handle failed payment event."""
    payment_id = data.get("id")
    error = data.get("last_error", {})
    error_code = error.get("code", "unknown")

    print(f"✗ Payment failed: {payment_id}")
    print(f"  Error: {error_code}")

    # TODO: Update order status
    # TODO: Notify customer


def handle_payment_canceled(data: dict) -> None:
    """Handle canceled payment event."""
    payment_id = data.get("id")
    print(f"⊘ Payment canceled: {payment_id}")

    # TODO: Update order status


def handle_refund_created(data: dict) -> None:
    """Handle refund created event."""
    refund_id = data.get("id")
    amount = data.get("amount")
    print(f"↩ Refund created: {refund_id} ({amount})")

    # TODO: Store refund record


def handle_refund_succeeded(data: dict) -> None:
    """Handle refund succeeded event."""
    refund_id = data.get("id")
    print(f"✓ Refund succeeded: {refund_id}")

    # TODO: Update refund status
    # TODO: Notify customer


def handle_payout_paid(data: dict) -> None:
    """Handle payout paid event."""
    payout_id = data.get("id")
    amount = data.get("amount")
    print(f"💸 Payout paid: {payout_id} ({amount})")

    # TODO: Update payout status
    # TODO: Notify merchant


def verify_and_process(signature_header: str, raw_body: bytes) -> bool:
    """
    Verify signature and process webhook.

    Args:
        signature_header: Molam-Signature header value
        raw_body: Raw request body

    Returns:
        True if verification and processing succeeded
    """
    try:
        # Verify signature
        print("Verifying signature...")
        is_valid = verify_signature(
            signature_header,
            raw_body,
            get_secret_by_kid,
            tolerance_ms=5 * 60 * 1000,  # 5 minutes
        )

        if not is_valid:
            print("✗ Signature verification failed")
            return False

        print("✓ Signature verified")

        # Parse and process event
        event = json.loads(raw_body.decode("utf-8"))
        process_webhook_event(event)

        return True

    except SignatureError as e:
        print(f"✗ Signature error: {e}")
        return False

    except json.JSONDecodeError as e:
        print(f"✗ Invalid JSON: {e}")
        return False

    except Exception as e:
        print(f"✗ Unexpected error: {e}")
        import traceback
        traceback.print_exc()
        return False


# ============================================================================
# Flask Webhook Endpoint
# ============================================================================

def create_flask_app():
    """
    Create Flask app with webhook endpoint.

    Usage:
        export MOLAM_WEBHOOK_SECRET="whsec_..."
        python examples/webhook_receiver.py --server
    """
    try:
        from flask import Flask, request, jsonify
    except ImportError:
        print("Error: Flask not installed. Run: pip install flask")
        sys.exit(1)

    app = Flask(__name__)

    @app.route("/webhook/molam", methods=["POST"])
    def molam_webhook():
        """Molam webhook endpoint."""
        signature = request.headers.get("Molam-Signature", "")
        raw_body = request.get_data()

        print(f"\n{'=' * 60}")
        print("Webhook received")
        print(f"{'=' * 60}")
        print(f"Signature: {signature[:50]}...")
        print(f"Body size: {len(raw_body)} bytes")

        if verify_and_process(signature, raw_body):
            return jsonify({"status": "ok"}), 200
        else:
            return jsonify({"error": "Verification failed"}), 401

    @app.route("/health", methods=["GET"])
    def health():
        """Health check endpoint."""
        return jsonify({"status": "healthy"}), 200

    return app


# ============================================================================
# CLI Interface
# ============================================================================

def main():
    """Main CLI interface."""
    if len(sys.argv) > 1 and sys.argv[1] == "--server":
        # Run Flask server
        print("Starting webhook server...")
        print("Webhook endpoint: http://localhost:5000/webhook/molam")
        print("Health check: http://localhost:5000/health")
        print("\nPress Ctrl+C to stop\n")

        app = create_flask_app()
        app.run(host="0.0.0.0", port=5000, debug=True)

    elif len(sys.argv) > 1:
        # Verify signature from command line
        signature_header = sys.argv[1]
        raw_body = sys.stdin.buffer.read()

        print(f"Signature: {signature_header}")
        print(f"Body: {raw_body.decode('utf-8')[:200]}...")
        print()

        if verify_and_process(signature_header, raw_body):
            print("\n✓ Verification successful")
            sys.exit(0)
        else:
            print("\n✗ Verification failed")
            sys.exit(1)

    else:
        print("Usage:")
        print("  # Start webhook server")
        print("  python examples/webhook_receiver.py --server")
        print()
        print("  # Verify signature from stdin")
        print("  echo '{...}' | python examples/webhook_receiver.py 't=...,v1=...,kid=1'")
        sys.exit(1)


if __name__ == "__main__":
    main()
